File: examples/checks/high_trust_bloom_envelope.py
# Independent Python checks for the high_trust_bloom_envelope example.
from __future__ import annotations

import math
import re

from .common import run_checks


def parse_answer(answer: str):
    def grab(pattern, cast=float):
        m = re.search(pattern, answer)
        return cast(m.group(1)) if m else None
    fp = re.search(r"false-positive envelope : ([0-9.]+) \.\. ([0-9.]+)", answer)
    return {
        "lambda": grab(r"lambda : ([0-9.]+)"),
        "fp_low": float(fp.group(1)) if fp else None,
        "fp_high": float(fp.group(2)) if fp else None,
        "extra": grab(r"expected extra exact lookups upper : ([0-9.]+)"),
    }


def run(ctx):
    data = ctx.load_input()
    a = data["Artifact"]
    reported = parse_answer(ctx.answer)
    lam = a["HashFunctions"] * a["CanonicalTripleCount"] / a["BloomBits"]
    fp_low = (1.0 - a["ExpMinusLambdaUpper"]) ** a["HashFunctions"]
    fp_high = (1.0 - a["ExpMinusLambdaLower"]) ** a["HashFunctions"]
    extra_upper = fp_high * a["NegativeLookupsPerBatch"]
    exact_exp = math.exp(-lam)

    checks = [
        ("numeric Bloom parameters are positive", all(a[k] > 0 for k in ["CanonicalTripleCount", "BloomBits", "HashFunctions", "NegativeLookupsPerBatch"])),
        ("canonical graph and SPO index triple counts agree", a["CanonicalTripleCount"] == a["SPOIndexTripleCount"]),
        ("lambda recomputes as k*n/m", abs(reported["lambda"] - lam) < 1e-12 and abs(lam - a["CertifiedLambda"]) < 1e-12),
        ("the exp(-lambda) decimal certificate brackets the exact Python value", a["ExpMinusLambdaLower"] <= exact_exp <= a["ExpMinusLambdaUpper"]),
        ("false-positive envelope is recomputed from the certificate", abs(reported["fp_low"] - fp_low) < 5e-10 and abs(reported["fp_high"] - fp_high) < 5e-10),
        ("false-positive upper bound stays below the policy budget", fp_high < a["FPRateBudget"]),
        ("expected extra exact lookups stay below budget", abs(reported["extra"] - extra_upper) < 5e-3 and extra_upper < a["ExtraExactLookupsBudget"]),
        ("maybe-positive answers must be confirmed against the canonical graph", data["Policies"]["MaybePositivePolicy"] == "ConfirmAgainstCanonicalGraph" and "maybe-positive policy : ConfirmAgainstCanonicalGraph" in ctx.answer),
        ("definite negatives may return absent without exact lookup", data["Policies"]["DefiniteNegativePolicy"] == "ReturnAbsent" and "definite-negative policy : ReturnAbsent" in ctx.answer),
        ("the deployment decision matches the recomputed envelope", data["Expected"]["Decision"] in ctx.answer and fp_high < a["FPRateBudget"] and extra_upper < a["ExtraExactLookupsBudget"]),
    ]
    return run_checks(checks)

File: examples/checks/high_trust_bloom_envelope.go
package main

import (
	"math"
)

func checkBloom(ctx *Context) []Check {
	d := ctx.M()
	a := asMap(d["Artifact"])
	lam := num(a["HashFunctions"]) * num(a["CanonicalTripleCount"]) / num(a["BloomBits"])
	fpLow := math.Pow(1-num(a["ExpMinusLambdaUpper"]), num(a["HashFunctions"]))
	fpHigh := math.Pow(1-num(a["ExpMinusLambdaLower"]), num(a["HashFunctions"]))
	extra := fpHigh * num(a["NegativeLookupsPerBatch"])
	exact := math.Exp(-lam)
	fp1, fp2 := math.NaN(), math.NaN()
	if m := reFind(ctx.Answer, `false-positive envelope : ([0-9.]+) \.\. ([0-9.]+)`); m != nil {
		fp1 = parseFloat(m[0])
		fp2 = parseFloat(m[1])
	}
	pos := true
	for _, k := range []string{"CanonicalTripleCount", "BloomBits", "HashFunctions", "NegativeLookupsPerBatch"} {
		pos = pos && num(a[k]) > 0
	}
	return []Check{{"numeric Bloom parameters are positive", pos}, {"canonical graph and SPO index triple counts agree", num(a["CanonicalTripleCount"]) == num(a["SPOIndexTripleCount"])}, {"lambda recomputes as k*n/m", close(fieldFloat(ctx.Answer, "lambda"), lam, 1e-12) && close(lam, num(a["CertifiedLambda"]), 1e-12)}, {"the exp(-lambda) decimal certificate brackets the exact value", num(a["ExpMinusLambdaLower"]) <= exact && exact <= num(a["ExpMinusLambdaUpper"])}, {"false-positive envelope is recomputed from the certificate", close(fp1, fpLow, 5e-10) && close(fp2, fpHigh, 5e-10)}, {"false-positive upper bound stays below the policy budget", fpHigh < num(a["FPRateBudget"])}, {"expected extra exact lookups stay below budget", close(fieldFloat(ctx.Answer, "expected extra exact lookups upper"), extra, 5e-3) && extra < num(a["ExtraExactLookupsBudget"])}, {"maybe-positive answers must be confirmed against the canonical graph", str(asMap(d["Policies"])["MaybePositivePolicy"]) == "ConfirmAgainstCanonicalGraph" && contains(ctx.Answer, "maybe-positive policy : ConfirmAgainstCanonicalGraph")}, {"definite negatives may return absent without exact lookup", str(asMap(d["Policies"])["DefiniteNegativePolicy"]) == "ReturnAbsent" && contains(ctx.Answer, "definite-negative policy : ReturnAbsent")}, {"the deployment decision matches the recomputed envelope", contains(ctx.Answer, str(asMap(d["Expected"])["Decision"])) && fpHigh < num(a["FPRateBudget"]) && extra < num(a["ExtraExactLookupsBudget"])}}
}
